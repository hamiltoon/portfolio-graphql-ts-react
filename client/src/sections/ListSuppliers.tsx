import React from 'react'
import { Loader, Table, Button, Message } from 'semantic-ui-react'
import { NavLink } from 'react-router-dom'

import { Supplier, RemoteData } from '../types'
import { routeMapping } from '../Routes'

type Props = { supplierData: RemoteData<Error, Supplier[]> }

const ListSuppliers = (props: Props) => {
  const { supplierData } = props

  switch (supplierData.type) {
    case 'NOT_ASKED':
      return null
    case 'LOADING':
      return (
        <Loader active inline="centered">
          Loading
        </Loader>
      )

    case 'SUCCESS':
      return (
        <Table celled>
          <Table.Header>
            <Table.Row>
              <Table.HeaderCell>Supplier id</Table.HeaderCell>
              <Table.HeaderCell>Name</Table.HeaderCell>
              <Table.HeaderCell>Created at</Table.HeaderCell>
              <Table.HeaderCell></Table.HeaderCell>
            </Table.Row>
          </Table.Header>

          <Table.Body>
            {supplierData.data.map((supplier) => (
              <Table.Row key={supplier.id}>
                <Table.Cell>{supplier.id}</Table.Cell>
                <Table.Cell>
                  <NavLink exact to={routeMapping.viewSuppliers.route(String(supplier.id))}>
                    {supplier.name}
                  </NavLink>
                </Table.Cell>
                <Table.Cell>{supplier.createdAt}</Table.Cell>
                <Table.Cell textAlign="right">
                  <NavLink exact to={routeMapping.editSuppliers.route(String(supplier.id))}>
                    <Button>Edit</Button>
                  </NavLink>
                </Table.Cell>
              </Table.Row>
            ))}
          </Table.Body>
        </Table>
      )
    case 'FAILURE':
      return (
        <Message>
          <Message.Header>Failed load the suppliers: </Message.Header>
          <p>{supplierData.error.message}</p>
        </Message>
      )
  }
}

export default ListSuppliers
