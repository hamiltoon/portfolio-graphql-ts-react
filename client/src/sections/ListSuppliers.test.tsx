import renderer from 'react-test-renderer'

import ListSuppliers from './ListSuppliers'

jest.mock('react-router-dom', () => ({
  NavLink: 'NavlinkMock',
}))

describe('Suppliers list', () => {
  test('before loading', () => {
    const componentToTest = renderer.create(<ListSuppliers supplierData={{ type: 'NOT_ASKED' }} />)
    expect(componentToTest).toMatchSnapshot()
  })
  test('loading', () => {
    const componentToTest = renderer.create(<ListSuppliers supplierData={{ type: 'LOADING' }} />)
    expect(componentToTest).toMatchSnapshot()
  })
  test('error', () => {
    const componentToTest = renderer.create(
      <ListSuppliers
        supplierData={{
          type: 'FAILURE',
          error: { name: 'some-error-name', message: 'some-message' },
        }}
      />,
    )
    expect(componentToTest).toMatchSnapshot()
  })
  test('success', () => {
    const componentToTest = renderer.create(
      <ListSuppliers
        supplierData={{
          type: 'SUCCESS',
          data: [
            {
              id: 1,
              name: 'some-supplier-name',
              createdAt: 'some-creation-date',
            },
          ],
        }}
      />,
    )
    expect(componentToTest).toMatchSnapshot()
  })
})
