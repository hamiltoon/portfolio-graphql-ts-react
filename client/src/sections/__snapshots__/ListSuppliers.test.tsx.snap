// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suppliers list before loading 1`] = `null`;

exports[`Suppliers list error 1`] = `
<div
  className="ui message"
>
  <div
    className="header"
  >
    Failed load the suppliers: 
  </div>
  <p>
    some-message
  </p>
</div>
`;

exports[`Suppliers list loading 1`] = `
<div
  className="ui active text centered inline loader"
>
  Loading
</div>
`;

exports[`Suppliers list success 1`] = `
<table
  className="ui celled table"
>
  <thead
    className=""
  >
    <tr
      className=""
    >
      <th
        className=""
      >
        Supplier id
      </th>
      <th
        className=""
      >
        Name
      </th>
      <th
        className=""
      >
        Created at
      </th>
      <th
        className=""
      />
    </tr>
  </thead>
  <tbody
    className=""
  >
    <tr
      className=""
    >
      <td
        className=""
      >
        1
      </td>
      <td
        className=""
      >
        <NavlinkMock
          exact={true}
          to="/suppliers/1"
        >
          some-supplier-name
        </NavlinkMock>
      </td>
      <td
        className=""
      >
        some-creation-date
      </td>
      <td
        className="right aligned"
      >
        <NavlinkMock
          exact={true}
          to="/suppliers/1/edit"
        >
          <button
            className="ui button"
            onClick={[Function]}
          >
            Edit
          </button>
        </NavlinkMock>
      </td>
    </tr>
  </tbody>
</table>
`;
